### ISSUE TITLE
Files/directories not found in /tmp/pip-install-trzvk4fh/pybloomfiltermmap3/pip-egg-info

### ISSUE THREAD

**Issue by zoeylee02** [prashnts/pybloomfiltermmap3#8]
It's Failed when execute `pip install  pybloomfiltermmap3 inside docker`
```
Collecting ecdsa==0.13 (from -r /home/docker/requirements.txt (line 8))
  Downloading https://files.pythonhosted.org/packages/63/f4/73669d51825516ce8c43b816c0a6b64cd6eb71d08b99820c00792cb42222/ecdsa-0.13-py2.py3-none-any.whl (86kB)
Collecting Cython==0.29.13 (from -r /home/docker/requirements.txt (line 9))
  Downloading https://files.pythonhosted.org/packages/45/f2/a7101b3457561e57f5abcd6f5ac13190054fecd7370f58f36fe2d6574742/Cython-0.29.13-cp36-cp36m-manylinux1_x86_64.whl (2.1MB)
Collecting pybloomfiltermmap3==0.4.17 (from -r /home/docker/requirements.txt (line 10))
  Downloading https://files.pythonhosted.org/packages/b8/c1/8b1c7b0db349e9a8d39cdc978e843c54302d6678adf096a32daa2a20987f/pybloomfiltermmap3-0.4.17.tar.gz (472kB)
ERROR: Files/directories not found in /tmp/pip-install-trzvk4fh/pybloomfiltermmap3/pip-egg-info
WARNING: You are using pip version 19.1.1, however version 19.2.3 is available.
You should consider upgrading via the 'pip install --upgrade pip' command.

```

**Comment by prashnts** [comment]
This could be very environment specific, and unfortunately I can't help you without more information. The base docker image you're building upon is a good place to start. :)

**Comment by mizvyt** [comment]
I found that this is dependent on Cython, meaning if pybloomfiltermmap3 is listed in requirements.txt, it will fail as requirements.txt does not necessarily preserve the order.

I believe that it is a bug that pybloomfiltermmap3 does not specify Cython as a dependency.

I added a PR to address this: https://github.com/prashnts/pybloomfiltermmap3/pulls

**Comment by prashnts** [comment]
Ok that definitely makes sense, thank you for more information. I will prep new release in morning. Thanks for the PR!

**Comment by prashnts** [comment]
Update: I got a bit of time now so I just pushed `v0.4.18` to pypi.

--
Closed in e31b690492d32cee72bda107f18cfc673ea6f18d

**Comment by mizvyt** [comment]
I have to wake this thread up..

I have looked closely at the code as it's not straightforward like this: the installation will still fail without Cython install as a prerequisite. I have made some changes in my fork to use a similar method as the Python 2 library uses: include the pybloomfilter.c file, and compile during setup if Cython is not available.

What is the reason that it was removed in the first place (as opposed to the Python 2 lib)? (ref commit 90caa8d)

ping @prashnts

**Comment by prashnts** [comment]
It's been quite long, but as far as I can remember, `pybloomfilter.c` is auto-generated as part of build process. See here: https://github.com/prashnts/pybloomfiltermmap3/blob/master/setup.py#L36 

Now that I think about it, it does seem you're right about having `cython` pre-installed. It should be possible to ensure cython is installed and implicitly install cython if otherwise, before the `setup()` method in `setup.py` executes. However, I don't think that's the best solution, since it may hide/replace cython specified in a dependency. Do you have an alternative solution that might be cleaner?

**Comment by mizvyt** [comment]
It seems that the usual way to go about it (used by other packages, such as [pandas](https://github.com/pandas-dev/pandas/blob/master/setup.py)) is to try using Cython to parse `.pyx`, otherwise fall back to compiling `.c`. The original Python 2 repo for this also uses a similar method.

To me it seems like it makes more sense to distribute a library with the compiled Cython code to avoid having to install additional dependencies and create workarounds for most users.

With such approach, it would probably be necessary to set up a hook for the development environment such that the `.pyx` gets compiled every time before integrating changes. Just my 2 cents in here.

**Comment by prashnts** [comment]
I am not sure if this isn't already the case: if you download the `source` from pypi (linked below), you can see that the `pybloomfilter.c` *is* included in the package. It's just not committed to git.

However, rest of your suggestion makes sense: we can simply fall back to building from `c` source instead of `pyx` source. The relevant change would be in `setup.py` here:

```
ext_files = [
  'src/mmapbitarray.c',
  'src/bloomfilter.c',
  'src/md5.c',
  'src/primetester.c',
  'src/MurmurHash3.c',
  'src/pybloomfilter.pyx'		<--
]
```

with cases where we'd test for cython or fall back to c. Mostly like in `pandas` you linked above (thanks!). We don't need the comprehensive templating like pandas anyway, so it should be fine.

> The original Python 2 repo for this also uses a similar method.

Pretty sure I probably changed that at some point. 😬 

I can't do this before next week however, would you have some time? :) 


**tl;dr**: `setup.py` needs to be improved so it's saner and less broken.

**Comment by mizvyt** [comment]
> I am not sure if this isn't already the case: if you download the `source` from pypi (linked below), you can see that the `pybloomfilter.c` _is_ included in the package. It's just not committed to git.

Interesting. I'll have a look.

> I can't do this before next week however, would you have some time? :)
>
> **tl;dr**: `setup.py` needs to be improved so it's saner and less broken.

I have modified it already to mostly match the other config, but I'll issue the PR tomorrow or the day after. I suggest we then review the approach, and extend / improve upon that.

**Comment by prashnts** [comment]
Forgot the link: https://files.pythonhosted.org/packages/5e/ba/8a5e20fd47a516045c8dc97ad0294ccac1e22ba1009df3112dec3ad1d617/pybloomfiltermmap3-0.4.18.tar.gz

And yes, that sounds great -- thank you!
